google-cloud-clients/google-cloud-firestore/src/test/java/com/google/cloud/firestore/it/ITSystemTest.java

```diff
@@ -662,6 +664,35 @@ public void listDocuments() throws Exception {
     assertEquals(documents.length, count);
   }
 
+  @Test
+  public void listDocumentsAsync() throws Exception {
+    // We test with 21 documents since 20 documents are by default returned in a single paged
+    // response.
+    String[] documents =
+        new String[] {
+          "1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12", "13", "14", "15", "16",
+          "17", "18", "19", "20", "21"
+        };
+    Arrays.sort(documents); // Sort in alphabetical (non-numeric) order.
+
+    WriteBatch batch = firestore.batch();
+    for (String document : documents) {
+      batch.create(randomColl.document(document), SINGLE_FIELD_OBJECT);
+    }
+    batch.commit().get();
+
+    ApiFuture<FirestoreClient.ListDocumentsPagedResponse> collectionRefs =
+        (ApiFuture<FirestoreClient.ListDocumentsPagedResponse>) randomColl.listDocumentAsync();
+    Iterator<Document> iterator = collectionRefs.get().iterateAll().iterator();
+
+    int count = 0;
+    for (Iterator<Document> documentIterator = iterator; documentIterator.hasNext(); ) {
```

Comment:
Unfortunately, this still blocks. iterateAll() uses an auto-paging iterator that receives the next page synchronously: https://github.com/googleapis/gax-java/blob/master/gax/src/main/java/com/google/api/gax/paging/AbstractPage.java#L118